### PR TITLE
Support a different elastic material in each block

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Getting code merged that's used in the thermal noise paper (https://arxiv.org/abs/2111.06893). This PR adds support for layered materials, but the DG scheme still needs an update to support this (next PR).

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: .clang-tidy [hunk 1]
```diff
@@ -78,6 +78,8 @@ Checks: '*,
 -readability-magic-numbers,
 # Same as llvm-qualified-auto above.,
 -readability-qualified-auto,
+# Access specifies can be useful to structure code,
```
Comment (geoffrey4444) [inline .clang-tidy]: specifies -> specifiers?

File: src/Domain/Creators/ExpandOverBlocks.hpp [hunk 2]
```diff
@@ -41,40 +67,151 @@ namespace domain {
...
+  }
 
  private:
   size_t num_blocks_;
   std::vector<std::string> block_names_;
   std::unordered_map<std::string, std::unordered_set<std::string>>
       block_groups_;
 };
+
+template <typename T>
```
Comment (nilsdeppe) [inline src/Domain/Creators/ExpandOverBlocks.hpp]: Could we move the definitions to a `tpp` and instantiate STL types maybe in `ExpandOverBlocks.cpp`? We can then also instantiate specific other types in other places as necessary. I'd us to continue to reduce code in header files. In this case I worry a bit that it'll be transitively included in a decent number of places.

File: src/Domain/Creators/ExpandOverBlocks.hpp [hunk 3]
```diff
@@ -41,40 +67,151 @@ namespace domain {
...
+std::vector<T> ExpandOverBlocks<T>::operator()(
+    const std::vector<T>& value) const {
+  if (value.size() != num_blocks_) {
+    throw std::length_error{"You supplied " + std::to_string(value.size()) +
+                            " values, but the domain creator has " +
+                            std::to_string(num_blocks_) + " blocks."};
+  }
+  if constexpr (tt::is_a_v<std::unique_ptr, T>) {
+    std::vector<T> expanded{};
```
Comment (nilsdeppe) [inline src/Domain/Creators/ExpandOverBlocks.hpp]: `Utilities/CloneUniquePtrs.hpp` has code to do this. Can you just use that?

File: src/Domain/Creators/ExpandOverBlocks.hpp [hunk 4]
```diff
@@ -41,40 +67,151 @@ namespace domain {
...
+template <typename T>
+std::vector<T> ExpandOverBlocks<T>::operator()(
+    const std::unordered_map<std::string, T>& value) const {
+  ASSERT(num_blocks_ == block_names_.size(),
+         "Construct 'ExpandOverBlocks' with block names to use the "
+         "map-over-block-names feature.");
+  // Expand group names
+  auto value_per_block = [&value]() {
+    if constexpr (tt::is_a_v<std::unique_ptr, T>) {
+      std::unordered_map<std::string, T> copy{};
```
Comment (nilsdeppe) [inline src/Domain/Creators/ExpandOverBlocks.hpp]: Again, can we use `Utilities/CloneUniquePtrs.hpp`?

File: src/ParallelAlgorithms/Events/ObserveNorms.hpp [hunk 5]
```diff
@@ -38,6 +38,7 @@
 #include "Utilities/ErrorHandling/Error.hpp"
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Events/ObserveNorms.hpp:38]: I don't understand how this change allows observing anything per material layer. It just allows you to name `ObserveNorms` something else. Couldn't this be achieved with compute tags filtering on the layer?

This needs tests regardless, but I'm not sure this is the right solution to the problem. Maybe we can discuss in the tech call?

File: src/Domain/Creators/BlockGroups.hpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * \brief Expand a list of block or group names into a list of block names
+ *
+ * \param block_or_group_names Block or group names to expand
+ * \param all_block_names All block names in the domain
+ * \param block_groups Block groups used to expand the names
+ * \return std::unordered_set<std::string> List of block names that appear in
+ * \p all_block_names. If one of the input names was a group, then all block
+ * names from that group are included. Overlaps between groups are allowed.
+ */
+std::unordered_set<std::string> expand_block_groups(
```
Comment (nilsdeppe) [inline src/Domain/Creators/BlockGroups.hpp]: maybe add `_to_block_names`? It isn't clear to me from the current name what it does, and while documented, the extra clarity seems worth it to me :shrug:

File: src/Elliptic/Systems/Elasticity/Actions/InitializeConstitutiveRelation.hpp [hunk 7]
```diff
@@ -0,0 +1,201 @@
+// Distributed under the MIT License.
```
Comment (nilsdeppe) [inline src/Elliptic/Systems/Elasticity/Actions/InitializeConstitutiveRelation.hpp:1]: Are there tests for these?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Just one quick thing I saw so far…haven’t finished reviewing yet

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Maybe I missed it, but is there a unit test anywhere that actually uses different constitutive relations in different blocks? If not, would it be hard to add one>?

File: src/ParallelAlgorithms/Events/ObserveNorms.hpp (hunk 5 above)
Comment (nilsvu) [reply src/ParallelAlgorithms/Events/ObserveNorms.hpp:38]: As discussed in a meeting recently, yes, this only provides a name for the event in the input file. It's needed when you have multiple `ObserveNorms` events in the same executable (here you have one for norms over the full domain and one for norms over each material layer). I think passing along a label struct is consistent with other parts of the code where we do that, e.g. `Options::Auto<T, Label>`. I added some docs to explain.

File: src/Domain/Creators/BlockGroups.hpp (hunk 6 above)
Comment (nilsvu) [reply src/Domain/Creators/BlockGroups.hpp]: squashed this rename in

**Comment by nilsvu** [comment]
I squashed in the split of ExpandOverBlocks into a tpp file, and added a test.

**Review by nilsdeppe — COMMENTED** [review]
LGTM, please squash!

**Comment by nilsvu** [comment]
@geoffrey4444 do you want to take another look?

**Review by geoffrey4444 — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**